src/lib/measure.c

```diff
@@ -0,0 +1,150 @@
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * Author: Slawomir Blauciak <slawomir.blauciak@linux.intel.com>
+ */
+
+#include <sof/measure.h>
+#include <sof/timer.h>
+
+struct measure_context measure_init(uint32_t average_steps)
+{
+	struct measure_context context;
+
+	memset_s(&context, sizeof(struct measure_context), 0,
+		 sizeof(struct measure_context));
+
+	context.steps = average_steps;
+	context.samples =
+		(struct measure_sample *)rzalloc(RZONE_RUNTIME,
+						 SOF_MEM_CAPS_RAM,
+						 average_steps *
+						 sizeof(struct measure_sample));
+
+	return context;
+}
+
+void measure_free(struct measure_context *context)
+{
+	rfree(context->samples);
+	context->samples = NULL;
+}
+
+static inline void measure_update_min_stat(struct measure_context *context)
+{
+	struct measure_sample *sample = &context->samples[context->cur_step];
+
+	if (sample->time_delta <
+	    context->stats[MEASURE_STAT_MIN].time_delta)
+		context->stats[MEASURE_STAT_MIN].time_delta =
+			sample->time_delta;
+
+	if (sample->ccount_delta <
+	    context->stats[MEASURE_STAT_MIN].ccount_delta)
+		context->stats[MEASURE_STAT_MIN].ccount_delta =
+			sample->ccount_delta;
+}
+
+static inline void measure_update_max_stat(struct measure_context *context)
+{
+	struct measure_sample *sample = &context->samples[context->cur_step];
+
+	if (sample->time_delta >
+	    context->stats[MEASURE_STAT_MAX].time_delta)
+		context->stats[MEASURE_STAT_MAX].time_delta =
+			sample->time_delta;
+
+	if (sample->ccount_delta >
+	    context->stats[MEASURE_STAT_MAX].ccount_delta)
+		context->stats[MEASURE_STAT_MAX].ccount_delta =
+			sample->ccount_delta;
+}
+
+static inline void measure_update_avg_stat(struct measure_context *context)
+{
+	int avg_step;
+
+	context->stats[MEASURE_STAT_AVG].ccount_delta = 0;
+	context->stats[MEASURE_STAT_AVG].time_delta = 0;
+
+	for (avg_step = 0; avg_step < context->steps; ++avg_step) {
+		context->stats[MEASURE_STAT_AVG].ccount_delta +=
+			context->samples[avg_step].ccount_delta;
+		context->stats[MEASURE_STAT_AVG].time_delta +=
+			context->samples[avg_step].time_delta;
+	}
+	context->stats[MEASURE_STAT_AVG].ccount_delta /=
+		context->steps;
+	context->stats[MEASURE_STAT_AVG].time_delta /=
+		context->steps;
+}
+
+static inline void measure_update_stats(struct measure_context *context)
+{
+	measure_update_min_stat(context);
+	measure_update_avg_stat(context);
+	measure_update_max_stat(context);
+}
+
+void measure_begin(struct measure_context *context)
+{
+	context->begin_time = platform_timer_get(platform_timer);
+
+	/* TODO: Handle 32-bit overflow of the CCOUNT register */
+	context->begin_ccount = xthal_get_ccount();
```

Comment:
Use timer API that wraps this xthal API.

Reply:
The problem that we don't support that timer API, since it relies on ccompare registers.
All hardware timers we use are external and therefore don't have respective ccompare registers.

Reply:
Not following you here, you are using CCOMPARE DSP register and appearing to compe to platform_timer_() (which can be async to CCOMPARE) on some platforms? CCOMPARE is also gated on some platforms during WFI.

Can you explain the relationship between bot time sources in the measurement ?

Reply:
There's no relationship between those two. I use both of them independently, precisely for the reason you mentioned: the fact that CCOUNT can be gated.

Reply:
ok, but you still haven't explained why you cant use timer api. xthal_get_ccount() is wrapped by timer API iirc.

Reply:
I believe I have. Our timer API requires internal hardware timers, which we don't use at all. The entire firmware relies on external TIMER3.

Reply:
Our timer APIs support both CAVS timers and DSP timers, both with 64 bit granularity. Please also see src/arch/xtensa/timer.c

Reply:
Yes, that's exactly the one I'm talking about. It uses the DSP timers (the CCOMPARE registers).

Reply:
yes if you need to, or register them in measure.c (using PLATFORM_ macros to identify the timers you want)